design/metadata-handling.md

```diff
@@ -0,0 +1,668 @@
+- **link**: The name of the link to configure this network for
+- **routes**: the list of route objects
+
+##### the services specifications
+
+The object for the **services** section list can be one of:
+
+- **dns**: a dns service with the ip address of a dns server
+
+### The Metal3Data object
+
+The output of the controller would be a Metal3Data object,one per node linking to the
+Metal3DataTemplate object and the associated secrets
+
+The Metal3Data object would be:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Data
+metadata:
+  name: nodepool-1-0
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3DataTemplate
+    name: nodepool-1
+spec:
+  index: 0
+  metaData:
+    name: machine-1-metadata
+    namespace: default
+  networkData:
+    name: machine-1-metadata
+    namespace: default
+  metal3Machine:
+    name: machine-1
+    namespace: default
+```
+The Metal3Data will contain the index of this node, and links to the secrets
+generated and to the Metal3Machine using this Metal3Data object.
+
+If the Metal3DataTemplate object is updated, the generated secrets will not be
+updated, to allow for reprovisioning of the nodes in the exact same state as
+they were initially provisioned. Hence, to do an update, it is necessary to do
+a rolling upgrade of all nodes.
+
+The reconciliation of the Metal3DataTemplate object will also be triggered by
+changes on Metal3Machines. In the case that a Metal3Machine gets modified, if
+the `dataTemplate` references a Metal3DataTemplate, that object will be reconciled.
+There will be two cases:
+
+- An already generated Metal3Data object exists with an ownerReference to this
+  Metal3Machine. In that case, the reconciler will verify that the required
+  secrets exist. If they do not, they will be created.
+- if no secret exists with an ownerReference to this Metal3Machine, then the
+  reconciler will create one and fill the respective field with the secret name.
+
+To create a Metal3Data object, the controller will select an index for that
+Metal3Machine. The selection happens by selecting the lowest available index.
+The controller will list all existing Metal3Data object linked to this
+Metal3DataTemplate and create a list of unavailable indexes. It will fill it by
+extracting the index from the Metal3Data names.
+
+Once the next available index is found, it will create the Metal3Data object.
+The name would be a concatenation of the Metal3DataTemplate name and index.
+Upon conflict, it will fetch again the list to consider the new list of
+Metal3Data and try to create the new object with the new index, this will happen
+until the new object is created successfully. Upon success, it will render the
+content values, and create the secrets containing the rendered data. The
+controller will generate the content based on the `metaData` or `networkData`
+field of the Metal3DataTemplate Specs.
+
```

Comment:
Let me see if I have the sequence right based on the description above.

For the fully automated flow, I think we have:

1. The user creates a Metal3DataTemplate (M3DT). It may contain any combination of metadata and network data.
2. The controller for M3DT creates a Metal3Data (M3D) with a reference to the template from 1, also giving it a unique index value and linking it to the Metal3Machine that will use it (how does it know that value?).
3. The M3D controller creates 2 Secrets and updates the spec.metaData and spec.networkData fields of the M3D to point to the Secrets. Either of the secrets may be empty, based on the data given. (Can you have an empty secret?)
4. The machine controller uses the secrets in the Metal3Machine to update the host as part of provisioning.

For the fully manual flow, I think we have:

1. The user creates two Secrets containing metadata and network data respectively.
2. The user creates a M3D with references to the secrets.
3. The user creates a Metal3Machine linked to the M3D from 2.
4. The machine controller uses the secrets in the Metal3Machine to update the host as part of provisioning.

For the partially automated flow, say where the user creates a metadata secret by hand:

1. The user creates one Secret containing metadata text.
2. The user creates a Metal3DataTemplate containing network data instructions.
3. The controller for M3DT creates an M3D with a reference to the secret from 1 and the template from 2, also giving it a unique index value and linking it to the Metal3Machine that will use it (how does it know that value?).
4. The M3D controller creates 1 Secret and updates the spec.networkData field of the M3D to point to the new Secret. It does not create a new Secret with metadata, regardless of the contents of the M3D, because the M3D already has a metadata secret set in its spec field.
5. The machine controller uses the secrets in the Metal3Machine to update the host as part of provisioning.

Reply:
I have added a whole section about the flow. About your questions : 
- the M3DT reconciler gets the M3Machine from the ownerreferences on the M3DT
- The secret would not be created if the `metaData` or `networkData` field is not set on the M3DT. in that case, the secret pointer is nil.
- for the partially automated flow, the user would give the secret directly in the Metal3Machine spec. In that case, the M3D controller will simply not create a `metaData` secret. If the `metaData` field is set on the M3DT, then it will also be created in the M3D, but not used in the M3Machine, because the `metaData` field in the spec is set.

Reply:
It sounds like the M3M controller will wait for the M3D to have any missing secret references, then copy those to the status fields of the M3M. And if the M3M spec includes references created by the user, those will be copied to the status instead. That way by looking at the status fields we can tell which secrets are actually being used by the M3M. Is that right?

Reply:
yes, this is correct.